File: src/server/usage.ts
/**
 * Usage click-logging (desktop side) — best-effort UI telemetry.
 *
 * Mirrors the launch ping (see license.ts `pingLaunch`): click events captured
 * in the renderer are buffered to a local JSONL ring-buffer under the vault
 * state dir, then batch-published to the hosted license Edge Function
 * (`op: "events"`) during startup — after the launch ping — and purged on
 * success (see license.ts `publishUsageEvents`). Nothing here may ever break a
 * launch: every path is best-effort and swallows its own errors.
 *
 * Consent: capture is OFF by default (opted out). The user must explicitly opt
 * in at registration (OnboardingPanel); the flag is persisted in `usage.json`
 * and a trial mint resets it to the opted-out default (license.ts `startTrial`),
 * so a new trial never silently re-enables capture. While opted out nothing is
 * captured, buffered, or published, and any buffered events are dropped.
 *
 * Privacy: only coarse labels are recorded (a control's text/aria-label or the
 * coarse folder/file kind) — never file/folder names, field values, file
 * contents, or secrets. Each label is length-capped on both ends of the wire.
 */
import fs from "node:fs";
import path from "node:path";
import { stateDir, readJson, writeJson } from "./config";
import { telemetryAllowed } from "./policy";

/** Coarse bucket for a captured interaction. Mirrors the renderer's capture hook. */
export type UsageEventType = "folder" | "file" | "toggle" | "button" | "link" | "nav" | "other";

const EVENT_TYPES: readonly UsageEventType[] = [
  "folder",
  "file",
  "toggle",
  "button",
  "link",
  "nav",
  "other",
];

export interface UsageEvent {
  /** ISO-8601 timestamp the interaction occurred. */
  at: string;
  type: UsageEventType;
  /** Coarse label/name of the touched control (no values/contents). */
  label: string;
}

// Ring-buffer caps — keep the MOST RECENT actions, trim the oldest on write.
const MAX_EVENTS = 5000;
const MAX_BYTES = 1_000_000; // ~1MB
const MAX_LABEL = 200; // labels are names, not content — clamp hard

const eventsPath = () => path.join(stateDir(), "usage-events.jsonl");
const consentPath = () => path.join(stateDir(), "usage.json");

interface UsageConsent {
  /** When true the user opted OUT — capture/buffer/publish are all disabled. */
  optOut?: boolean;
}

/** Whether the user has opted OUT of usage logging. Default is opted OUT:
 *  capture stays off until the user explicitly opts in (optOut === false).
 *  A managed `telemetry: "off"` policy reads as permanently opted out —
 *  this one gate locks capture (append), publish, and the UI toggle state. */
export function isUsageOptedOut(): boolean {
  if (!telemetryAllowed()) return true;
  return readJson<UsageConsent>(consentPath(), {}).optOut !== false;
}

/**
 * Persist the consent flag. Opting out also drops any buffered-but-unpublished
 * events, so nothing already captured leaks out on the next launch.
 */
export function setUsageOptOut(optOut: boolean): void {
  writeJson(consentPath(), { optOut } satisfies UsageConsent);
  if (optOut) clearUsageBuffer();
}

/** Reset consent to the default (opted OUT). Called when a trial is minted so a
 *  new trial never silently re-enables capture; the user's explicit choice is
 *  persisted afterward by the onboarding flow. */
export function resetUsageConsent(): void {
  writeJson(consentPath(), { optOut: true } satisfies UsageConsent);
}

/** Drop the local buffer (best-effort). */
function clearUsageBuffer(): void {
  try {
    fs.rmSync(eventsPath(), { force: true });
  } catch {
    /* best-effort */
  }
}

/** Write text atomically (temp-then-rename) to avoid torn reads. */
let writeCounter = 0;
function writeTextAtomic(file: string, text: string): void {
  const tmp = `${file}.${process.pid}.${writeCounter++}.tmp`;
  fs.writeFileSync(tmp, text, "utf8");
  fs.renameSync(tmp, file);
}

/** Coerce one untrusted event from the renderer, or null if unusable. */
function sanitize(raw: unknown): UsageEvent | null {
  if (!raw || typeof raw !== "object") return null;
  const e = raw as Record<string, unknown>;
  const type = EVENT_TYPES.includes(e.type as UsageEventType) ? (e.type as UsageEventType) : "other";
  const label = typeof e.label === "string" ? e.label.replace(/\s+/g, " ").trim().slice(0, MAX_LABEL) : "";
  if (!label) return null;
  const at = typeof e.at === "string" && !Number.isNaN(Date.parse(e.at)) ? e.at : new Date().toISOString();
  return { at, type, label };
}

/**
 * Append captured events to the local buffer, then trim to the ring-buffer cap
 * (keep the most recent by count and by size). No-op when opted out. Tolerates
 * a pre-existing partial last line by reparsing what it can.
 */
export function appendUsageEvents(events: unknown[]): void {
  if (isUsageOptedOut()) return;
  const incoming = events.map(sanitize).filter((e): e is UsageEvent => e !== null);
  if (!incoming.length) return;
  try {
    const existing = readBufferLines();
    let lines = [...existing, ...incoming.map((e) => JSON.stringify(e))];
    if (lines.length > MAX_EVENTS) lines = lines.slice(lines.length - MAX_EVENTS);
    // Enforce the byte cap by dropping the oldest lines until under the limit.
    let text = lines.join("\n") + "\n";
    while (text.length > MAX_BYTES && lines.length > 1) {
      lines = lines.slice(1);
      text = lines.join("\n") + "\n";
    }
    writeTextAtomic(eventsPath(), text);
  } catch {
    /* a full buffer is best-effort — never surface to the caller */
  }
}

/** Read the raw buffer as lines (empty/partial lines dropped). */
function readBufferLines(): string[] {
  try {
    return fs
      .readFileSync(eventsPath(), "utf8")
      .split("\n")
      .map((l) => l.trim())
      .filter(Boolean)
      .filter((l) => {
        try {
          JSON.parse(l);
          return true;
        } catch {
          return false; // tolerate a torn/partial line
        }
      });
  } catch {
    return []; // missing buffer
  }
}

/** The parsed buffer plus the line count, so a successful publish can purge it. */
export function readUsageBuffer(): { events: UsageEvent[]; lineCount: number } {
  const lines = readBufferLines();
  const events = lines
    .map((l) => {
      try {
        return sanitize(JSON.parse(l));
      } catch {
        return null;
      }
    })
    .filter((e): e is UsageEvent => e !== null);
  return { events, lineCount: lines.length };
}

/**
 * Purge the first `publishedLineCount` lines (the oldest, just published),
 * preserving any events appended since the read. New events only ever append at
 * the end, so dropping the leading lines is safe.
 */
export function purgeUsageBuffer(publishedLineCount: number): void {
  if (publishedLineCount <= 0) return;
  try {
    const lines = readBufferLines();
    const remaining = lines.slice(publishedLineCount);
    if (remaining.length === 0) clearUsageBuffer();
    else writeTextAtomic(eventsPath(), remaining.join("\n") + "\n");
  } catch {
    /* best-effort */
  }
}

File: native/crates/lighthouse-core/src/usage.rs
//! Usage click-logging (port of `src/server/usage.ts`) — best-effort UI
//! telemetry. Consent is OFF by default (opted out); only coarse labels are
//! recorded, ring-buffered to `usage-events.jsonl` and batch-published on
//! launch. Nothing here may ever break a launch.

use std::fs;
use std::path::PathBuf;
use std::sync::atomic::{AtomicU64, Ordering};

use serde::{Deserialize, Serialize};

use crate::config::{iso_now, read_json, state_dir, write_json};

const EVENT_TYPES: &[&str] = &["folder", "file", "toggle", "button", "link", "nav", "other"];

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UsageEvent {
    pub at: String,
    #[serde(rename = "type")]
    pub event_type: String,
    pub label: String,
}

// Ring-buffer caps — keep the MOST RECENT actions, trim the oldest on write.
const MAX_EVENTS: usize = 5000;
const MAX_BYTES: usize = 1_000_000;
const MAX_LABEL: usize = 200;

fn events_path() -> PathBuf {
    state_dir().join("usage-events.jsonl")
}
fn consent_path() -> PathBuf {
    state_dir().join("usage.json")
}

#[derive(Debug, Default, Serialize, Deserialize)]
struct UsageConsent {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    opt_out: Option<bool>,
    // The TS file writes camelCase `optOut`; accept both spellings on read.
    #[serde(default, rename = "optOut", skip_serializing_if = "Option::is_none")]
    opt_out_camel: Option<bool>,
}

impl UsageConsent {
    fn value(&self) -> Option<bool> {
        self.opt_out_camel.or(self.opt_out)
    }
}

/// Whether the user has opted OUT of usage logging. Default is opted OUT.
/// A managed `telemetry: "off"` policy reads as permanently opted out —
/// this one gate locks capture (append), publish, and the UI toggle state.
pub fn is_usage_opted_out() -> bool {
    if !crate::policy::telemetry_allowed() {
        return true;
    }
    read_json(&consent_path(), UsageConsent::default()).value() != Some(false)
}

/// Persist the consent flag. Opting out also drops any buffered events.
pub fn set_usage_opt_out(opt_out: bool) {
    write_json(&consent_path(), &serde_json::json!({ "optOut": opt_out }));
    if opt_out {
        clear_usage_buffer();
    }
}

/// Reset consent to the default (opted OUT) — called when a trial is minted.
pub fn reset_usage_consent() {
    write_json(&consent_path(), &serde_json::json!({ "optOut": true }));
}

fn clear_usage_buffer() {
    let _ = fs::remove_file(events_path());
}

static WRITE_COUNTER: AtomicU64 = AtomicU64::new(0);

/// Write text atomically (temp-then-rename) to avoid torn reads.
fn write_text_atomic(file: &PathBuf, text: &str) {
    let n = WRITE_COUNTER.fetch_add(1, Ordering::Relaxed);
    let tmp = file.with_file_name(format!(
        "{}.{}.{}.tmp",
        file.file_name().and_then(|s| s.to_str()).unwrap_or("usage"),
        std::process::id(),
        n
    ));
    if fs::write(&tmp, text).is_ok() {
        let _ = fs::rename(&tmp, file);
    } else {
        let _ = fs::remove_file(&tmp);
    }
}

/// Coerce one untrusted event from the renderer, or None if unusable.
fn sanitize(raw: &serde_json::Value) -> Option<UsageEvent> {
    let obj = raw.as_object()?;
    let event_type = obj
        .get("type")
        .and_then(|t| t.as_str())
        .filter(|t| EVENT_TYPES.contains(t))
        .unwrap_or("other")
        .to_string();
    let label = obj
        .get("label")
        .and_then(|l| l.as_str())
        .map(|l| {
            let collapsed: String = l.split_whitespace().collect::<Vec<_>>().join(" ");
            collapsed.chars().take(MAX_LABEL).collect::<String>()
        })
        .unwrap_or_default();
    if label.is_empty() {
        return None;
    }
    let at = obj
        .get("at")
        .and_then(|a| a.as_str())
        .filter(|a| crate::config::parse_ms(a).is_some())
        .map(String::from)
        .unwrap_or_else(iso_now);
    Some(UsageEvent {
        at,
        event_type,
        label,
    })
}

/// Read the raw buffer as lines (empty/partial lines dropped).
fn read_buffer_lines() -> Vec<String> {
    let Ok(text) = fs::read_to_string(events_path()) else {
        return Vec::new();
    };
    text.split('\n')
        .map(str::trim)
        .filter(|l| !l.is_empty())
        .filter(|l| serde_json::from_str::<serde_json::Value>(l).is_ok())
        .map(String::from)
        .collect()
}

/// Append captured events to the local buffer, then trim to the ring caps.
/// No-op when opted out.
pub fn append_usage_events(events: &[serde_json::Value]) {
    if is_usage_opted_out() {
        return;
    }
    let incoming: Vec<String> = events
        .iter()
        .filter_map(sanitize)
        .filter_map(|e| serde_json::to_string(&e).ok())
        .collect();
    if incoming.is_empty() {
        return;
    }
    let mut lines = read_buffer_lines();
    lines.extend(incoming);
    if lines.len() > MAX_EVENTS {
        lines = lines.split_off(lines.len() - MAX_EVENTS);
    }
    // Enforce the byte cap by dropping the oldest lines until under the limit.
    let mut text = lines.join("\n") + "\n";
    while text.len() > MAX_BYTES && lines.len() > 1 {
        lines.remove(0);
        text = lines.join("\n") + "\n";
    }
    write_text_atomic(&events_path(), &text);
}

/// The parsed buffer plus the line count, so a successful publish can purge it.
pub fn read_usage_buffer() -> (Vec<UsageEvent>, usize) {
    let lines = read_buffer_lines();
    let events = lines
        .iter()
        .filter_map(|l| serde_json::from_str::<serde_json::Value>(l).ok())
        .filter_map(|v| sanitize(&v))
        .collect();
    (events, lines.len())
}

/// Purge the first `published_line_count` lines (the oldest, just published),
/// preserving any events appended since the read.
pub fn purge_usage_buffer(published_line_count: usize) {
    if published_line_count == 0 {
        return;
    }
    let lines = read_buffer_lines();
    let remaining: Vec<String> = lines.into_iter().skip(published_line_count).collect();
    if remaining.is_empty() {
        clear_usage_buffer();
    } else {
        write_text_atomic(&events_path(), &(remaining.join("\n") + "\n"));
    }
}
